### PR TITLE
fix: using tombstones to account for rapid deletion

### PR DESCRIPTION
**Pull request by shawkins**
closes: #2314

If during a creation the events for both the add and the deletion occur before the resource is being added to the temporary cache, the temporary cache moves backwards in a way that may not resolve itself.

The approach here is to use tombstones to track deletions. The benefit of this approach is that it keeps the logic encapsulated to the TemporaryResourceCache. The drawback is that we're only guessing about how long to retain the tombstones.

The other approach would look more similar to event recording, which is notify the cache that a create is happening prior to when it starts so that deletions / tombstones can gathered for that specific resource.

### CODE REVIEW

File: operator-framework-core/src/main/java/io/javaoperatorsdk/operator/processing/event/source/informer/TemporaryResourceCache.java [hunk 1]
```diff
@@ -60,7 +63,23 @@ protected boolean removeEldestEntry(java.util.Map.Entry<String, Boolean> eldest)
...
+  }
+
+  public synchronized void onAddOrUpdateEvent(T resource) {
+    onEvent(resource, false);
+  }
+
+  synchronized void onEvent(T resource, boolean unknownState) {
+    if (!tombstones.isEmpty()) {
+      var iter = tombstones.entrySet().iterator();
+      var entry = iter.next();
```
Comment (csviri) [inline operator-framework-core/src/main/java/io/javaoperatorsdk/operator/processing/event/source/informer/TemporaryResourceCache.java:78]: Not sure if understand this part, it always removes one random event from the tombstone if expired? Didn't you mean to iterate over all events?

File: operator-framework-core/src/main/java/io/javaoperatorsdk/operator/processing/event/source/informer/TemporaryResourceCache.java [hunk 2]
```diff
@@ -38,8 +38,11 @@ public class TemporaryResourceCache<T extends HasMetadata> {
 
   private static final Logger log = LoggerFactory.getLogger(TemporaryResourceCache.class);
   private static final int MAX_RESOURCE_VERSIONS = 256;
+  private static final int TOMBSTONE_RETENTION_TIME = 120000;
```
Comment (csviri) [inline operator-framework-core/src/main/java/io/javaoperatorsdk/operator/processing/event/source/informer/TemporaryResourceCache.java:41]: Wondering about this value, if high enough.

File: operator-framework-core/src/main/java/io/javaoperatorsdk/operator/processing/event/source/informer/TemporaryResourceCache.java [hunk 3]
```diff
@@ -84,14 +103,25 @@ public synchronized void putResource(T newResource, String previousResourceVersi
...
-    if ((previousResourceVersion == null && cachedResource == null)
+    boolean moveAhead = false;
+    if (previousResourceVersion == null && cachedResource == null) {
+      if (tombstones.get(newResource.getMetadata().getUid()) != null) {
+        log.debug(
+            "Won't resurrect uid {} for resource id: {}",
+            newResource.getMetadata().getUid(), resourceId);
+        return;
+      }
+      moveAhead = true;
```
Comment (csviri) [inline operator-framework-core/src/main/java/io/javaoperatorsdk/operator/processing/event/source/informer/TemporaryResourceCache.java:114]: THis seems to be correct, but would be good to document why is this here, isn't is covered already by the condition on line 120-121 ? or just an optimization?

**Review by csviri — CHANGES_REQUESTED** [review]

File: operator-framework-core/src/main/java/io/javaoperatorsdk/operator/processing/event/source/informer/TemporaryResourceCache.java (hunk 1 above)
Comment (shawkins) [reply operator-framework-core/src/main/java/io/javaoperatorsdk/operator/processing/event/source/informer/TemporaryResourceCache.java:78]: No, this is to amortorize the expiration logic - you could of course decide on a greater number than 1.

File: operator-framework-core/src/main/java/io/javaoperatorsdk/operator/processing/event/source/informer/TemporaryResourceCache.java (hunk 2 above)
Comment (shawkins) [reply operator-framework-core/src/main/java/io/javaoperatorsdk/operator/processing/event/source/informer/TemporaryResourceCache.java:41]: It is certianly possible there could be an extreme scenario where the client is configured for a bunch of exponential retries - I'm not sure there is a hard cap on the amount of time in that case. We can keep things for longer, or you can go back the other direction of telling the cache prior to the create that you are intending to create the given resource.

File: operator-framework-core/src/main/java/io/javaoperatorsdk/operator/processing/event/source/informer/TemporaryResourceCache.java (hunk 3 above)
Comment (shawkins) [reply operator-framework-core/src/main/java/io/javaoperatorsdk/operator/processing/event/source/informer/TemporaryResourceCache.java:114]: I'll add a comment, it's just breaking apart the checks that were being performed.

File: operator-framework-core/src/main/java/io/javaoperatorsdk/operator/processing/event/source/informer/TemporaryResourceCache.java (hunk 1 above)
Comment (csviri) [reply operator-framework-core/src/main/java/io/javaoperatorsdk/operator/processing/event/source/informer/TemporaryResourceCache.java:78]: Maybe desribe in there as a comment too, that is intentional, since I will at least naivly stop on this every time I open it :)

File: operator-framework-core/src/main/java/io/javaoperatorsdk/operator/processing/event/source/informer/TemporaryResourceCache.java (hunk 3 above)
Comment (csviri) [reply operator-framework-core/src/main/java/io/javaoperatorsdk/operator/processing/event/source/informer/TemporaryResourceCache.java:114]: thank you!

File: operator-framework-core/src/main/java/io/javaoperatorsdk/operator/processing/event/source/informer/TemporaryResourceCache.java (hunk 1 above)
Comment (metacosm) [reply operator-framework-core/src/main/java/io/javaoperatorsdk/operator/processing/event/source/informer/TemporaryResourceCache.java:78]: I would like to see this as a comment, indeed, as well.

**Review by metacosm — CHANGES_REQUESTED** [review]

**Review by metacosm — CHANGES_REQUESTED** [review]
LGTM apart from missing comment explaining the amortization of cache emptying logic

**Comment by shawkins** [comment]
@metacosm @metacosm updated the code with further comments and consolidated the secondary caching concerns to a single class. From what I can tell it will take around 170 MB to track up to 1000000 deletions. Also updated the base retention time to 10 minutes, rather than 2. Based upon seeing what the retained heap size of this could be, I also updated the knownResourceVersions cache to be able to store more entries.

**Review by metacosm — APPROVED** [review]
